docs/coding-guidelines/vectorization-guidelines.md

```diff
@@ -0,0 +1,1091 @@
+
+### Access violation (AV) testing
+
+Handling the remainder in an invalid way may lead to non-deterministic and hard to diagnose issues.
+
+Let's look at the following code:
+
+```diff
+nuint elementOffset = 0;
+while (elementOffset < (nuint)buffer.Length)
+{
+    loaded = Vector128.LoadUnsafe(ref searchSpace, elementOffset); // BUG!
+
+    elementOffset += (nuint)Vector128<int>.Count;
+}
+```
+
+How many times will the loop execute for a buffer of six integers? Twice! The first time it will load the first four elements, but the second time it will load the random content of the memory following the buffer!
+
+Writing tests that detect that issue is hard, but not impossible. The .NET Team uses a helper utility called [BoundedMemory](https://github.com/dotnet/runtime/blob/main/src/libraries/Common/tests/TestUtilities/System/Buffers/BoundedMemory.Creation.cs) that allocates a memory region which is immediately preceded by or immediately followed by a poison (`MEM_NOACCESS`) page. Attempting to read the memory immediately before or after it results in `AccessViolationException`.
+
+## Loading and storing vectors
+
+### Loading
+
+Both `Vector128` and `Vector256` provide at least five ways of loading them from memory:
+
+```cs
+public static class Vector128
+{
+    public static Vector128<T> Load<T>(T* source) where T : unmanaged;
+    public static Vector128<T> LoadAligned<T>(T* source) where T : unmanaged;
+    public static Vector128<T> LoadAlignedNonTemporal<T>(T* source) where T : unmanaged;
+    public static Vector128<T> LoadUnsafe<T>(ref T source) where T : struct;
+    public static Vector128<T> LoadUnsafe<T>(ref T source, nuint elementOffset) where T : struct;
+}
+```
+
+The first three overloads require a pointer to the source. To be able to use a pointer to a managed buffer in a safe way, the buffer needs to be pinned first. This is because the GC cannot track unmanaged pointers. It needs help to ensure that it doesn't move the memory while you're using it, as the pointers would silently become invalid. The tricky part here is doing the pointer arithmetic right:
+
+```cs
+unsafe int UnmanagedPointersSum(Span<int> buffer)
+{
+    fixed (int* pBuffer = buffer)
+    {
+        int* pEnd = pBuffer + buffer.Length;
+        int* pOneVectorFromEnd = pEnd - Vector128<int>.Count;
+        int* pCurrent = pBuffer;
+
+        Vector128<int> sum = Vector128<int>.Zero;
+
+        while (pCurrent <= pOneVectorFromEnd)
+        {
+            sum += Vector128.Load(pCurrent);
+
+            pCurrent += Vector128<int>.Count;
+        }
+
+        int result = Vector128.Sum(sum);
+
+        while (pCurrent < pEnd)
+        {
+            result += *pCurrent;
+
+            pCurrent++;
+        }
+
+        return result;
+    }
+}
+```
+
+`LoadAligned` and `LoadAlignedNonTemporal` require the input to be aligned. Aligned reads and writes should be slightly faster but using them comes at a price of increased complexity.
```

Comment:
Might be worth calling out that `NonTemporal` means that the hardware is allowed to bypass the cache. This can help provide a speedup when working with very large amounts of data as it avoids repeatedly filling the cache with values that will never be used again.